client.go

```diff
@@ -376,14 +376,28 @@ func (c *client) periodicQuery(ctx context.Context, params *lookupParams) error
 	bo.MaxInterval = 60 * time.Second
 	bo.Reset()
 
+	var timer *time.Timer
+	defer func() {
+		if timer != nil {
+			timer.Stop()
+		}
+	}()
 	for {
 		// Backoff and cancel logic.
 		wait := bo.NextBackOff()
 		if wait == backoff.Stop {
 			return fmt.Errorf("periodicQuery: abort due to timeout")
 		}
+		if timer == nil {
+			timer = time.NewTimer(wait)
+		} else {
+			if !timer.Stop() {
+				<-timer.C
```

Comment:
Thank you for subscribing to "facts about timers in go"!

This can deadlock. `timer.Stop()` will return true iff it stops the timer. If the timer already fired and we've already read off the channel in line 400, `timer.Stop()` will return `false` (timer already fired) and this read will block forever.

Reply:
Fixed.